ftw/solr/upgrades/20200401111512_update_modified_index/upgrade.py

```diff
@@ -0,0 +1,163 @@
+    value = to_iso8601(date)
+    if value.endswith('.000Z'):
+        value = value[:-5] + 'Z'
+    return value
+
+
+def _pre_datetime_format_fix_solr_date(date):
+    """Solr date representation. Fractional seconds are stripped if 0."""
+    value = _pre_datetime_format_fix_to_iso8601(date)
+    if value.endswith('.000Z'):
+        value = value[:-5] + 'Z'
+    return value
+
+
+commit_interval = 100
+
+
+class UpdateModifiedIndex(UpgradeStep):
+    """Update modified index.
+    """
+
+    def __call__(self):
+        self.logger = logging.getLogger('ftw.upgrade')
+
+        if not self.is_enabled():
+            self.log("Skipping: Solr is not enabled.")
+            return
+        self.sync()
+
+    def is_enabled(self):
+        registry = queryUtility(IRegistry)
+        settings = registry.forInterface(ISolrSettings)
+        return settings.enabled
+
+    def get_objects_that_need_syncing_of_modified(self):
+        """Find objects that are not in sync only due to change in DateTime
+        rounding"""
+
+        catalog = getToolByName(self.portal, 'portal_catalog')
+        items = catalog.unrestrictedSearchResults()
+        catalog_modified = set(
+            [(item.UID, _solr_date(item.modified)) for item in items])
+        pre_datetime_format_fix_catalog_modified = set(
+            [(item.UID, _pre_datetime_format_fix_solr_date(item.modified)) for item in items])
+
+        conn = self.manager.connection
+        resp = conn.search({
+            u'query': u'*:*',
+            u'limit': 10000000,
+            u'params': {u'fl': ['UID', 'modified']},
+        })
+        solr_modified = set(
+            [(doc['UID'], doc.get('modified', u'2000-01-01T00:00:00.000Z'))
+             for doc in resp.docs])
+
+        not_in_sync = set(
+            (item[0] for item in catalog_modified - solr_modified))
+        pre_datetime_format_fix_not_in_sync = set(
+            (item[0] for item in pre_datetime_format_fix_catalog_modified - solr_modified))
+
+        # We only want to reindex modified for objects that would
+        # be in sync if we hadn't modified the rounding in solr_date
+        to_correct = not_in_sync - pre_datetime_format_fix_not_in_sync
+        return to_correct
+
+    def sync(self, commit_interval=100, idxs=None, doom=True):
+        """Sync Solr with portal catalog"""
+        if not self.is_enabled():
+            return 'Solr indexing is disabled.'
+
+        to_correct = self.get_objects_that_need_syncing_of_modified()
+
+        # Reindex modified for Solr items that should be in sync
+        processed = 0
+        real = timer()
+        lap = timer()
+        cpu = timer(clock)
+
+        if doom:
+            transaction.doom()
+
+        zodb_conn = self.portal._p_jar
+
+        def commit():
+            conn = self.manager.connection
+            conn.commit(soft_commit=False, extract_after_commit=False)
+            zodb_conn.cacheGC()
```

Comment:
This might be counterproductive. The peak of cache usage is already reached by calling `get_objects_that_need_syncing_of_modified()`. After that, all brains are in the cache. There should be no significant cache growth during processing of `to_correct` and thus no need to garbage collect. If the cache is really garbage collected, then brains might need to be loaded again from the database.

Reply:
Right thanks. I copied that without much thinking.